src/thor/pathalgorithm.cc

```diff
@@ -0,0 +1,128 @@
+#include "thor/pathalgorithm.h"
+#include "baldr/datetime.h"
+#include "midgard/logging.h"
+#include "sif/edgelabel.h"
+
+#include <chrono>
+#include <ctime>
+
+namespace dt = valhalla::baldr::DateTime;
+namespace sc = std::chrono;
+
+namespace valhalla {
+namespace thor {
+
+// helper function to initialize the object from a location
+TimeInfo TimeInfo::make(valhalla::Location& location, baldr::GraphReader& reader) {
+  // No time to to track
+  if (!location.has_date_time()) {
+    return {false};
+  }
+
+  // Get the timezone of the first edge's end node that we can
+  int timezone_index = 0;
+  for (const auto& pe : location.path_edges()) {
+    const baldr::GraphTile* tile = nullptr;
+    const auto* edge = reader.directededge(baldr::GraphId(pe.graph_id()), tile);
+    timezone_index = reader.GetTimezone(edge->endnode(), tile);
+    if (timezone_index != 0)
+      break;
+  }
+
+  // Set the origin timezone to be the timezone at the end node use this for timezone changes
+  if (timezone_index == 0) {
+    LOG_WARN("No timezone for location using default");
+    timezone_index = 1;
+  }
+  const auto* tz = dt::get_tz_db().from_index(timezone_index);
+
+  // Get the current local time and set it if its a current time route
+  // NOTE: we set to minute resolution to match the input format
+  const auto now_date =
+      date::make_zoned(tz, sc::time_point_cast<sc::seconds>(
+                               sc::time_point_cast<sc::minutes>(sc::system_clock::now())));
+  if (location.date_time() == "current") {
+    std::ostringstream iso_dt;
+    iso_dt << date::format("%FT%R", now_date);
+    location.set_date_time(iso_dt.str());
+  }
+
+  // Convert the requested time into seconds from epoch
+  date::local_seconds parsed_date;
+  try {
+    parsed_date = dt::get_formatted_date(location.date_time(), true);
+  } catch (...) {
+    LOG_ERROR("Could not parse provided date_time: " + location.date_time());
+    return {false};
+  }
+  const auto then_date = date::make_zoned(tz, parsed_date);
+  uint64_t local_time = date::to_utc_time(then_date.get_sys_time()).time_since_epoch().count();
+
+  // What second of the week is this (for historical traffic lookup)
+  std::tm t = dt::iso_to_tm(location.date_time());
+  std::mktime(&t);
+  auto second_of_week = t.tm_wday * valhalla::midgard::kSecondsPerDay +
+                        t.tm_hour * valhalla::midgard::kSecondsPerHour + t.tm_sec;
+
+  // When is this route with respect to now this will let us appropriately use current flow traffic
+  int64_t seconds_from_now = (then_date.get_local_time() - now_date.get_local_time()).count();
+
+  // Make a valid object
+  return {true, timezone_index, local_time, second_of_week, seconds_from_now};
+}
+
+// offset all the initial time info to reflect the progress along the route to this point
+TimeInfo TimeInfo::operator+(Offset offset) const {
```

Comment:
this is for forward expansion and offsets from the original time info by the offset provided. the offset is a number of seconds (from costing) and a target timezone (to do timezone offsetting of localtime and second of week).